share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,162 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+    def edit f j =
+        require (JInteger ji) = j else j
+        JInteger (f ji)
+
+    edit
+
+export def jBoolean: JValue => List Boolean =
+    def view j =
+        require (JBoolean jb) = j else Nil
+        jb, Nil
+
+    view
+
+export def jeBoolean: (Boolean => Boolean) => JValue => JValue =
+    def edit f j =
+        require (JBoolean jb) = j else j
+        JBoolean (f jb)
+
+    edit
+
+# Are these safe since 1.0 => JDouble 1.0 => to_json => from_json => JInteger 1?
+# Maybe jNumber is the only *safe* thing to use? On the other hand I'm worried
+# about only giving jNumber as an abstraction over jDouble...IDK yet.
+export def jDouble: JValue => List Double =
+    def view j =
+        require (JDouble jd) = j else Nil
+        jd, Nil
+
+    view
+
+export def jeDouble: (Double => Double) => JValue => JValue =
+    def edit f j =
+        require (JDouble jd) = j else j
+        JDouble (f jd)
+
+    edit
+
+export def jNumber: JValue => List Double =
+    def view = match _
+        JDouble x  = x, Nil
+        JInteger x = fromOption $ dint x
+        _          = Nil
+
+    view
+
+# This is odd because if you edit a JInteger with say +1 you still get a JDouble
+export def jeNumber: (Double => Double) => JValue => JValue =
+    def edit f s = match s
+        JDouble x  = JDouble (f x)
+        JInteger x =
+            require Some dx = dint x else s
+            JDouble (f dx)
+        _          = s
+
+    edit
+
+export def jNull: JValue => List Unit =
+    def view j =
+        require JNull = j else Nil
+        Unit, Nil
+
+    view
+
+export def jeNull: (Unit => Unit) => JValue => JValue =
+    def edit f j =
+        require JNull = j else j
+        def _ = f Unit # the edit function might be effectful
+        JNull
+
+    edit
+
+export def jObject: JValue => List (List (Pair String JValue)) =
+    def view j =
+        require (JObject jo) = j else Nil
+        jo, Nil
+
+    view
+
+export def jeObject: (List (Pair String JValue) => List (Pair String JValue)) => JValue => JValue =
+    def edit f j =
+        require (JObject jo) = j else j
+        JObject (f jo)
+
+    edit
+
+export def jArray: JValue => List (List JValue) =
+    def view j =
+        require (JArray ja) = j else Nil
+        ja, Nil
+
+    view
+
+export def jeArray: (List JValue => List JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray jl) = j else j
+        JArray (f jl)
+
+    edit
+
+export def jSingelton: JValue => List JValue =
+    def view j =
+        require (JArray (j2, Nil)) = j else Nil
+        j2, Nil
+
+    view
+
+export def jeSingelton: (JValue => JValue) => JValue => JValue =
+    def edit f j =
+        require (JArray (j2, Nil)) = j else j
+        JArray (f j2, Nil)
+
+    edit
+
+export def jAll: JValue => List JValue = _ | jArray |> (_)
+export def jeAll: (JValue => JValue) => JValue => JValue = jeArray $ map $ _
+
+export def jAt (i: Integer): JValue => List JValue = _ | jArray |> atT i
```

Comment:
... `= jAll _ | at i | optionToList` or `= jAll _ | drop i | take 1`?

Reply:
I'm fine with `jAll _ | at i | optionToList`

We had talked about changing `at` to return a List but I think both of us might not want to do that. optionToList in core solves that thought I suppose. I wonder if we can be more ergonomic?

Reply:
Yeah, we can't change `at`. I think we want list operations to remain focussed on being ergonomic in non-json contexts. If they happen to compose with minimal effort (I'd say `| optionToList` is minimal effort), that's gravy.